src/google/protobuf/compiler/command_line_interface.cc

```diff
@@ -188,8 +188,25 @@ bool TryCreateParentDirectory(const string& prefix, const string& filename) {
 // Get the absolute path of this protoc binary.
 bool GetProtocAbsolutePath(string* path) {
 #ifdef _WIN32
-  char buffer[MAX_PATH];
-  int len = GetModuleFileName(NULL, buffer, MAX_PATH);
+  wchar_t wide_buffer[MAX_PATH];
+  // Note: the returned length does not include terminating null character.
+  int len = GetModuleFileNameW(NULL, wide_buffer, MAX_PATH);
+  if (len == GOOGLE_ARRAYSIZE(wide_buffer) &&
+      GetLastError() == ERROR_INSUFFICIENT_BUFFER) {
+    // The path is too long.
+    return false;
+  }
+  char buffer[MAX_PATH + 1];
+  for (size_t i = 0; i < len; ++i) {
+    wchar_t wide_char = wide_buffer[i];
+    if (wide_char > 0x7F) {
+      // non-ASCII character is not supported yet.
+      // TODO(yukawa): Support paths that contain non-ASCII characters.
```

Comment:
> If there is an encoding with which "/includle" is not a valid byte sequence, protocol compiler will definitely not work on such platforms no matter what we do here (e.g., we replace ".proto" with ".pb.h" when generating header files. If ".pb.h" is invalid, ...)

Sorry, my example was misleading. Please disregard my comment that `"/includle"` could be invalid.  However, `+ "/include"` is still problematic because there are some character encodings that are stateful. [ISO/IEC 2022](https://en.wikipedia.org/wiki/ISO/IEC_2022#Code_structure) is an example.  If we want to concatenate two strings whose encoding is ISO/IEC 2022, we may or may not need to insert state change characters (e.g. `0x0F` or `0x0E`) to change shift state depending on the state of each string.

Having said that, my original goal of this CL is to fix the build breakage, which is actually the same to the goal of the internal CL that @TeBoring mentioned.  Dealing with an arbitrary system locale in the runtime environment is beyond the goal of this CL. Hence I'm OK with the following change if you are happy.
```
  char buffer[MAX_PATH];
  int len = GetModuleFileNameA(NULL, buffer, MAX_PATH);
```

> Protocol compiler uses lower-level file APIs to read and write files (e.g., open or fopen).

I understood this point.

> I believe these APIs does not handle character encodings and you have to use char sequences in the system locale to successfully open files with non-ASCII names.

I agree that we have to use char sequences in the system locale because of the above reason. However, correctly manipulating any MBCS is indeed difficult and I'm afraid that our current has already missed to cover some edge cases.  Fixing that is not the goal of this CL though.

Thanks,